source/develop/release-management/features/virt/dedicated-cpu.md

```diff
@@ -0,0 +1,170 @@
+---
+title: Dedicated CPU Policies
+category: feature
+authors: tgolembi
+---
+
+# Dedicated CPU Policies
+
+## Summary
+
+This feature introduces policies for dynamically dedicating physical CPUs to
+the VM. While this can be (to some extent) done with static CPU pinning, the
+CPU pinning requires that the VM is pinned also to a particular host and that
+no other VMs (without pinning) are running on the host. Instead, this new
+feature allows the user to configure only the way how should physical CPUs be
+assigned to the VM and oVirt will take care of dynamically allocating suitable
+physical CPUs as well as evicting other VMs from those CPUs.
+
+
+## Owner
+
+* Name: Tomáš Golembiovský
+* Email: tgolembi@redhat.com
+
+
+## Description
+
+To be able to have more flexibility in configure how virtual CPUs (vCPUs)
+of VMs use physical CPUs (pCPUs) of the host this feature introduces several
+policies for dedicating CPUs. Compared to CPU pinning the advantages are that:
+
+  * VM does not need to be pinned to a specific host,
+  * solves some of the issues with *existing* pinning policy where
+    several VMs can end up to be pinned to same CPUs,
+  * *TODO* I am pretty sure there were couple more worth mentioning here
+
+The available policies described below divide the set of available pCPUs into
+three sub-sets, There are the dedicated CPUs that are assigned only to a single
+VM. Then there are some unassigned CPUs that cannot be used by any other VM
+because of restrictions requirements on the threads of the core. And finally
+there is the set of CPUs that are not strictly assigned to any particular VM
+and are shared between all the VMs with *shared* policy. This subset is hereby
+referred to as *shared pool*.
+
+The available CPU policies are:
+
+  * *shared*: *(default)* The guest vCPUs will be allowed to roam among all
+    pCPUs from shared pool.
+  * *dedicated*: The guest vCPUs will be strictly pinned to a set of host pCPUs
+    (similarly to static CPU pinning). The set of pCPUs will be chosen to match
+    the required guest CPU topology. If the host has SMT architecture, thread
+    siblings are preferred.
+  * *isolate-threads*: If the host does not have SMT architecture this is same
+    as *dedicated* policy. If the host has SMT architecture each vCPU is placed
+    on a different physical core and all thread siblings are excluded from use
+    by this and any other VM (rendering all but one thread from a core unusable).
```

Comment:
Does it mean there is no space for a policy that would isolate threads in the described sense, but without strictly pinning guest vCPUs to pCPUs?

Reply:
Let's say you request a VM with 3 vCPUs and the host has 2 threads per core. Do you mean each vCPU should be able to move between 2 (threads in core) pCPUs or all 6 (threds times cores) pCPUs? Either way, I don't see what could be benefit of that (performance or otherwise).

Reply:
No, I confused static pinning with dedicated pinning here, sorry. I feel like the document would benefit from a short glossary at its beginning.

Reply:
I had the same problem as Milan with confusion on a offline discussion. Pinning != Dedicated and it should be highlighted.

About the isolate-threads itself, I think 1 thread == 1 vCPU, if a host has 2 threads per core and the VM has 2 vCPUs it should take 1 core with the 2 threads in it. Is it talking about 3 vCPUs when one core will be split? Why not allowing VM with 1 vCPU to use that one free thread? Or, let it be shareable?
If this is limiting it, what is the reason for this policy? 
Same for the bottom threads limitation for `require-threads`.

Reply:
> I had the same problem as Milan with confusion on a offline discussion. Pinning != Dedicated and it should be highlighted.

Ok, I am adding a short glossary in a next version.

> 
> About the isolate-threads itself, I think 1 thread == 1 vCPU, if a host has 2 threads per core and the VM has 2 vCPUs it should take 1 core with the 2 threads in it.

No, `isolate-threads` will place exactly 1 vCPU on a core, no matter how many threads are in the core. On the other hand `require-threads` does what you are describing. It will fit as many vCPUs on a core as there are threads.

> Is it talking about 3 vCPUs when one core will be split? Why not allowing VM with 1 vCPU to use that one free thread? Or, let it be shareable?
> If this is limiting it, what is the reason for this policy?
> Same for the bottom threads limitation for `require-threads`.

The point of `isolate-threads` and `require-threads` policy is to make sure no other VMs share the cores in use.

And although we could let the vCPUs float around threads in the core if there are unused threads, it brings up a question how to schedule the vCPUs. What if there are multiple vCPUs on the core? Let's say you have 4 threads per core and you only need to use 2, will one vCPU be pinned to 1 thread and the other vCPU pinned to the other 3 or will each vCPU have 2 threads? What if you have 3 vCPUs on such core? Which one of the three vCPUs will be the winner with 2 threads? 

All in all I believe that letting the vCPUs roam around the core defeats the purpose of dedicating CPUs. Of course, the counterargument is that the scheduler on host probably knows best what to do.